### DISCUSSION TITLE
Create Attachment with Bitwarden CLI

### DISCUSSION THREAD

**Discussion by rbreunung** [dani-garcia/vaultwarden#4217] [Q&A] [answered]
### Subject of the issue
Error on creating attachment using Bitwarden CLI on Windows / Mac.

### Your environment (Generated via diagnostics page)
* Vaultwarden version: v1.30.1
* Web-vault version: v2023.10.0
* OS/Arch: linux/x86_64
* Running within Docker: true (Base: Alpine)
* Environment settings overridden: true
* Uses a reverse proxy: true
* IP Header check: false (X-Forwarded-For)
* Internet access: true
* Internet access via a proxy: false
* DNS Check: true
* Browser/Server Time Check: true
* Server/NTP Time Check: true
* Domain Configuration Check: true
* HTTPS Check: true
* Database type: SQLite
* Database version: 3.44.0
* Clients used: 
* Reverse proxy and version: 
* Other relevant information: 

### Config (Generated via diagnostics page)
<details><summary>Show Running Config</summary>

**Environment settings which are overridden:** DOMAIN, SIGNUPS_ALLOWED, SIGNUPS_VERIFY, ADMIN_TOKEN

```json
{
  "_duo_akey": null,
  "_enable_duo": false,
  "_enable_email_2fa": true,
  "_enable_smtp": true,
  "_enable_yubico": false,
  "_icon_service_csp": "",
  "_icon_service_url": "",
  "_ip_header_enabled": true,
  "_smtp_img_src": "cid:",
  "admin_ratelimit_max_burst": 3,
  "admin_ratelimit_seconds": 300,
  "admin_session_lifetime": 20,
  "admin_token": "***",
  "allowed_iframe_ancestors": "",
  "attachments_folder": "data/attachments",
  "auth_request_purge_schedule": "30 * * * * *",
  "authenticator_disable_time_drift": false,
  "data_folder": "data",
  "database_conn_init": "",
  "database_max_conns": 10,
  "database_timeout": 30,
  "database_url": "***************",
  "db_connection_retries": 15,
  "disable_2fa_remember": false,
  "disable_admin_token": false,
  "disable_icon_download": false,
  "domain": "*****://********************",
  "domain_origin": "*****://********************",
  "domain_path": "",
  "domain_set": true,
  "duo_host": null,
  "duo_ikey": null,
  "duo_skey": null,
  "email_attempts_limit": 3,
  "email_change_allowed": true,
  "email_expiration_time": 600,
  "email_token_size": 6,
  "emergency_access_allowed": true,
  "emergency_notification_reminder_schedule": "0 3 * * * *",
  "emergency_request_timeout_schedule": "0 7 * * * *",
  "enable_db_wal": true,
  "event_cleanup_schedule": "0 10 0 * * *",
  "events_days_retain": null,
  "extended_logging": true,
  "helo_name": null,
  "hibp_api_key": null,
  "icon_blacklist_non_global_ips": true,
  "icon_blacklist_regex": null,
  "icon_cache_folder": "data/icon_cache",
  "icon_cache_negttl": 259200,
  "icon_cache_ttl": 2592000,
  "icon_download_timeout": 10,
  "icon_redirect_code": 302,
  "icon_service": "internal",
  "incomplete_2fa_schedule": "30 * * * * *",
  "incomplete_2fa_time_limit": 3,
  "invitation_expiration_hours": 120,
  "invitation_org_name": "Vaultwarden",
  "invitations_allowed": true,
  "ip_header": "X-Real-IP",
  "job_poll_interval_ms": 30000,
  "log_file": "/logs/vaultwarden.log",
  "log_level": "warn",
  "log_timestamp_format": "%Y-%m-%d %H:%M:%S.%3f",
  "login_ratelimit_max_burst": 10,
  "login_ratelimit_seconds": 60,
  "org_attachment_limit": null,
  "org_creation_users": "",
  "org_events_enabled": false,
  "org_groups_enabled": false,
  "password_hints_allowed": true,
  "password_iterations": 600000,
  "push_enabled": false,
  "push_installation_id": "***",
  "push_installation_key": "***",
  "push_relay_uri": "https://push.bitwarden.com",
  "reload_templates": false,
  "require_device_email": false,
  "rsa_key_filename": "data/rsa_key",
  "send_purge_schedule": "0 5 * * * *",
  "sendmail_command": null,
  "sends_allowed": true,
  "sends_folder": "data/sends",
  "show_password_hint": false,
  "signups_allowed": false,
  "signups_domains_whitelist": "",
  "signups_verify": true,
  "signups_verify_resend_limit": 6,
  "signups_verify_resend_time": 3600,
  "smtp_accept_invalid_certs": false,
  "smtp_accept_invalid_hostnames": false,
  "smtp_auth_mechanism": null,
  "smtp_debug": false,
  "smtp_embed_images": true,
  "smtp_explicit_tls": null,
  "smtp_from": "**********************",
  "smtp_from_name": "Vaultwarden",
  "smtp_host": "****************",
  "smtp_password": "***",
  "smtp_port": 587,
  "smtp_security": "starttls",
  "smtp_ssl": null,
  "smtp_timeout": 15,
  "smtp_username": "**********************",
  "templates_folder": "data/templates",
  "tmp_folder": "data/tmp",
  "trash_auto_delete_days": null,
  "trash_purge_schedule": "0 5 0 * * *",
  "use_sendmail": false,
  "use_syslog": false,
  "user_attachment_limit": null,
  "web_vault_enabled": true,
  "web_vault_folder": "web-vault/",
  "websocket_address": "0.0.0.0",
  "websocket_enabled": false,
  "websocket_port": 3012,
  "yubico_client_id": null,
  "yubico_secret_key": null,
  "yubico_server": null
}
```
</details>

* vaultwarden version: 1.30.1

* Install method: community container of Nextcloud AIO

* Clients used: Bitwarden CLI 2023.12.1 2023.9.0 on Windows 11 and Mac Sonoma (latest patch level)

* Reverse proxy and version: 

* MySQL/MariaDB or PostgreSQL version: 

* Other relevant details:

Basically, I installed Nextcloud AIO as recommended using Docker on Ubuntu. On top of this I installed a community container setup of Caddy Fail2Ban and VaultWarden following the recommendations. The Web Vault and UI clients work for me as far as I tested.

### Steps to reproduce

- have a VaultWarden account with an item in a folder.
- know the item id and have an existing file on your computer for upload
- have the Bitwarden CLI correctly logged in to your VaultWarden account
- Try to create an attachment using the CLI e.g.
```powershell
bw create attachment --file ./Backup-codes.txt --itemid 36423883-7d1e-4aba-aa03-f27838c71217
```

### Expected behaviour

- attachment is created in the vault
- return value of the CLI is a JSON

### Actual behaviour

- Bitwarden CLI response

```html
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="utf-8">
    <meta name="color-scheme" content="light dark">
    <title>400 Bad Request</title>
</head>
<body align="center">
    <div role="main" align="center">
        <h1>400: Bad Request</h1>
        <p>The request could not be understood by the server due to malformed syntax.</p>
        <hr />
    </div>
    <div role="contentinfo" align="center">
        <small>Rocket</small>
    </div>
</body>
</html>
```

### Troubleshooting data

- I also tested against my BitWarden account. Here the command is working as expected.
- I tested with my PC and my Mac. The Bitwarden CLI is creating the same result.
- Web Vault and Desktop App create attachments as expected in VaultWarden.
- I want to use the CLI as part of a Powershell script to export and re-import a complete vault including all attachment as I understand this is a missing feature of the official Bitwarden client. You can find the project in my GitHub.

**Comment by BlackDex** [comment]
See #3934

**Reply by rbreunung** [reply]
Thank you for the quick reply. 

Do you have any insight it may be related to specific Node.js versions?

**Reply by BlackDex** [reply, marked as answer]
It works with 16, it breaks with 18.
It also is a know bug at Bitwarden.

https://github.com/bitwarden/clients/issues/5618
https://github.com/bitwarden/clients/issues/5876

**Reply by mrclschstr** [reply]
I am experiencing similar problems when exporting/downloading attachments with the Bitwarden CLI (`bw get attachment`). I have installed vaultwarden 1.30.5 and use the CLI 2024.2.1 under Windows 10 (installation with chocolatey).

I am very sure I am using the correct parameters but only get a "Not found" (and nothing more) back from the CLI. Are the problems somehow related?

**Reply by mrclschstr** [reply]
Well I solved it... the documentation is wrong. For more details see: https://community.bitwarden.com/t/download-attachment-fails-when-using-the-cli/51462/3

**Comment by rbreunung** [comment]
Thank you very much! I can confirm, when switching to Node.js 16 using `nvm`, then everything is working as expected.

**Comment by rbreunung** [comment]
Dear all,

as I try to continue developing my backup script I re-evaluated current situation. I tested on Windows 11 and with nodejs 16 / 18 / 20 via nvm. 

<details>

<summary>Details</summary>

- bw CLI is 2024.2.0

In this example command I properly exchanged relevant UUIDs.

```powershell
bw create attachment --file '.\export-2024-05-20-20-28-37\2ccce09e-fd8a-46ce-a193-b073013e68a6\id_rsa' --itemid 912b92dc-e154-4e66-97cc-0cb8db6e31db
```

</details>

<details>

<summary>Error Message</summary>

```html
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="utf-8">
    <meta name="color-scheme" content="light dark">
    <title>400 Bad Request</title>
</head>
<body align="center">
    <div role="main" align="center">
        <h1>400: Bad Request</h1>
        <p>The request could not be understood by the server due to malformed syntax.</p>
        <hr />
    </div>
    <div role="contentinfo" align="center">
        <small>Rocket</small>
    </div>
</body>
</html>
```

</details>

Result:

- Using **official vault.bitwarden.com** server the attachment upload **worked**. I switched to Node 16 and 20 and they both worked as expected.
- Using latest **Vaultwarden** Docker image with Vaultwarden 1.30.5 results in the mentioned **error** message for both Node 16 and 20. 
- Using web interfaces of both servers works like expected. 

This means the issue is fixed for Bitwarden but somehow remains in Vaultwarden. There is also no solution by using an older Node version.

Do you have any hints or a fix for me?

**Reply by rbreunung** [reply]
I have one addition to my post. I think I did a mistake by using the CLI .exe via winget. I removed in and installed it via NPM on node 16, 18,20.

Unfortunately the CLI is no more compatible with Note16 but breaking with following error on every call.

```log
D:\antrophos\nvm\v16.20.2\node_modules\@bitwarden\cli\build\bw.js:19740
        const toStore = structuredClone(obj);
                        ^

ReferenceError: structuredClone is not defined
    at MemoryStorageService.save (D:\antrophos\nvm\v16.20.2\node_modules\@bitwarden\cli\build\bw.js:19740:25)
    at StateService.<anonymous> (D:\antrophos\nvm\v16.20.2\node_modules\@bitwarden\cli\build\bw.js:26141:45)
    at Generator.next (<anonymous>)
    at D:\antrophos\nvm\v16.20.2\node_modules\@bitwarden\cli\build\bw.js:25053:71
    at new Promise (<anonymous>)
    at state_service_awaiter (D:\antrophos\nvm\v16.20.2\node_modules\@bitwarden\cli\build\bw.js:25049:12)
    at StateService.setState (D:\antrophos\nvm\v16.20.2\node_modules\@bitwarden\cli\build\bw.js:26140:16)
    at StateService.<anonymous> (D:\antrophos\nvm\v16.20.2\node_modules\@bitwarden\cli\build\bw.js:25120:32)
    at Generator.next (<anonymous>)
    at D:\antrophos\nvm\v16.20.2\node_modules\@bitwarden\cli\build\bw.js:25053:71
```

Wouldn't it be an option to support the broken header in the server, as there is no improvement expected?

**Reply by BlackDex** [reply]
No, we can't support this in a sane way. The client just doesn't send data in a valid way that we can extract using the http server we use.

The issue is also still open at the Bitwarden repo https://github.com/bitwarden/clients/issues/5618
And it looks like they seem to have created a PR which seems to solve this issue https://github.com/bitwarden/clients/pull/7342

**Comment by V1rgul** [comment]
For people looking for a temporary solution, confirming it works with
- Server : `vaultwarden/server:1.30.5` (latest as of now) docker image
- Client
  - Node.js 16.20.2 (latest 16) runtime
  - `@bitwarden/cli@2023.4.0` prebuilt npm package

Using a newer nodejs or a newer cli client doesn't

**Reply by rbreunung** [reply]
This is a good inspiration. Maybe set up some package.json for the project.

**Comment by querylab** [comment]
I am experiencing the exact same problem with Vaultwarden. It looks like the problem is on Bitwarden side, as mentioned in pull request [bitwarden/clients#7342 ](https://github.com/bitwarden/clients/pull/7342)
